R/binary_classification.R

```diff
@@ -138,3 +138,210 @@ fbeta_score <- function(actual, predicted, beta = 1) {
     return((1 + beta^2) * prec * rec / ((beta^2 * prec) + rec))
 }
+
+#' Binary confusion matrix
+#'
+#' \code{cmat} Calculates a binary classification confusion matrix,
+#' comparing the predicted with the actual values for the classes.
+#' Assumes that 1 is used for the positive class and 0 for the
+#' negative class.
+#' Returns a \code{data.frame} with columns corresponding to the
+#' number of True Positives (\code{tp}), False Positives (\code{fp}),
+#' True Negatives (\code{tn}), and False Negatives (\code{fn})
+#'
+#' @inheritParams params_binary
+#' @seealso \code{\link{sensitivity}} \code{\link{specificity}}
+#' @examples
+#' actual <- c(1, 1, 1, 0, 0, 0)
+#' predicted <- c(1, 0, 1, 1, 1, 1)
+#' cmat(actual, predicted)
+
+cmat <- function(actual, predicted) {
+    tbl <- as.data.frame(table(actual, predicted))
+    cm <- data.frame(tp = NA, fn = NA, fp = NA, tn = NA)
+    cm[, c("tn", "fp", "fn", "tp")] <- tbl$Freq
+    cm
+}
+
+#' Sensitivity
+#'
+#' \code{sensitivity} calculates the proportion of actual positives
+#' that are correctly identified as such. It is also known as
+#' \code{true positive rate}.
+#'
+#' @inheritParams params_binary
+#' @seealso \code{\link{cmat}} \code{\link{specificity}}
+#' @export
+#' @examples
+#' actual <- c(1, 1, 1, 0, 0, 0)
+#' predicted <- c(1, 0, 1, 1, 1, 1)
+#' sensitivity(actual, predicted)
+
+sensitivity <- function(actual, predicted) {
+    cm <- cmat(actual, predicted)
+    cm$tp / (cm$tp + cm$fn)
+}
+
+#' Specificity
+#'
+#' \code{specificity} calculates the proportion of actual negatives
+#' that are correctly identified as such. It is also known as
+#' \code{true negative rate}.
+#'
+#' @inheritParams params_binary
+#' @seealso \code{\link{cmat}} \code{\link{sensitivity}}
+#' @export
+#' @examples
+#' actual <- c(1, 1, 1, 0, 0, 0)
+#' predicted <- c(1, 0, 1, 1, 1, 1)
+#' specificity(actual, predicted)
+
+specificity <- function(actual, predicted) {
+    cm <- cmat(actual, predicted)
+    cm$tn / (cm$tn + cm$fp)
+}
+
+#' False Negative Rate
+#'
+#' \code{fnr} calculates the proportion of actual positives
+#' that are not identified as such.
+#' It is defined as \code{1 - sensitivity}
+#'
+#' @inheritParams params_binary
+#' @export
+#' @seealso \code{\link{cmat}} \code{\link{sensitivity}}
+#' @examples
+#' actual <- c(1, 1, 1, 0, 0, 0)
+#' predicted <- c(1, 0, 1, 1, 1, 1)
+#' fnr(actual, predicted)
+
+fnr <- function(actual, predicted) {
+    1 - sensitivity(actual, predicted)
+}
+
+#' False Positive Rate
+#'
+#' \code{fnr} calculates the proportion of actual negatives
+#' that are not identified as such.
+#' It is defined as \code{1 - specificity}
+#'
+#' @inheritParams params_binary
+#' @seealso \code{\link{cmat}} \code{\link{specificity}}
+#' @export
+#' @examples
+#' actual <- c(1, 1, 1, 0, 0, 0)
+#' predicted <- c(1, 0, 1, 1, 1, 1)
+#' fpr(actual, predicted)
+
+fpr <- function(actual, predicted) {
+    1 - specificity(actual, predicted)
+}
+
+#' Positive Predictive Value
```

Comment:
Similar to `sensitivity`, this metric is equivalent to `recall`, which is already defined in this package. Can you update this function so that it uses that function?

Reply:
Actually, PPV = TP / (TP + FP), wheres recall = sensitivity = TP / (TP + FN), so they are not measuring the same concept

Reply:
ppv is formally equivalent to precision, made it refer to the implementation of the latter, which now uses confusion_matrix